### PR TITLE
stats: Predefine symbols needed for user agent stats

### PR DESCRIPTION
**Pull request by jmarantz**
Description: Envoy automatically and unconditionally tracks some properties related to user-agent if the request-header is present. This PR predeclares the symbols needed to avoid taking locks at runtime. This had previously flown under the radar because the call to Scope::counter() did not match the syntax in the check_format scan. That syntax has now been generalized but there was still a workaround. This is now solved with a proper context, owned by Http::Context. 
Risk Level: low
Testing: //test/...
Docs Changes: n/a
Release Notes: n/a

### CODE REVIEW

**Comment by htuch** [comment]
(not sure why API shepherds were tagged, going to assume this was due to some earlier patch that is no longer present)

**Comment by htuch** [comment]
@itayd do you know what's up with the LGTM failure ^^?

**Comment by itayd** [comment]
@htuch i don't see any api files in here? might be a badly phrased error then: https://github.com/repokitteh/modules/blob/master/ownerscheck.star#L208

wdyt?

**Comment by jmarantz** [comment]
I think something went wrong in a merge, and suddenly a bunch of files I
didn't touch were part of the PR. Then I did another merge and it fixed
itself.

File: source/common/http/user_agent.h [hunk 1]
```diff
@@ -10,29 +10,41 @@
...
-  COUNTER(downstream_cx_destroy_remote_active_rq)                                                  \
-  COUNTER(downstream_rq_total)
+namespace Http {
 
-/**
- * Wrapper struct for user agent stats. @see stats_macros.h
- */
-struct UserAgentStats {
-  ALL_USER_AGENTS_STATS(GENERATE_COUNTER_STRUCT)
+struct UserAgentContext {
```
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/common/http/user_agent.h]: I added some comments to the classes; basically the Context is common to all scopes, and the Stats structure is specific to a scope.

File: source/common/http/conn_manager_config.h [hunk 2]
```diff
@@ -79,8 +80,18 @@ struct ConnectionManagerNamedStats {
...
+      : named_(std::move(named_stats)), prefix_(prefix),
+        prefix_stat_name_storage_(
+            std::make_unique<Stats::StatNameManagedStorage>(prefix, scope.symbolTable())),
+        scope_(scope) {}
+
+  Stats::StatName prefixStatName() const { return prefix_stat_name_storage_->statName(); }
+
   ConnectionManagerNamedStats named_;
   std::string prefix_;
+  std::unique_ptr<Stats::StatNameManagedStorage> prefix_stat_name_storage_;
```
[the remark this answers is not in the document]
Comment (jmarantz) [reply source/common/http/conn_manager_config.h]: It did, but I removed the need by adding a std::move constructor to StatNameManagedStorage; there was already one for its parent class so it was pretty trivial and this simplifies the code. Good call.

File: source/common/http/context_impl.h [hunk 3]
```diff
@@ -3,6 +3,9 @@
 #include "envoy/http/context.h"
 
 #include "common/http/codes.h"
+#include "common/http/user_agent.h"
+
+//#include "common/tracing/http_tracer_impl.h"
```
Comment (snowp) [inline source/common/http/context_impl.h]: remove

File: source/extensions/filters/network/http_connection_manager/config.h [hunk 4]
```diff
@@ -20,6 +20,7 @@
 #include "common/json/json_loader.h"
 #include "common/router/rds_impl.h"
 #include "common/router/scoped_rds.h"
+#include "common/stats/symbol_table_impl.h"
```
Comment (snowp) [inline source/extensions/filters/network/http_connection_manager/config.h]: this can probably also be reverted?

**Pull request merged**